lib/Optimizer/TrainingPreparation.cpp

```diff
@@ -0,0 +1,147 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "glow/Optimizer/TrainingPreparation.h"
+#include "glow/Graph/PlaceholderBindings.h"
+
+namespace glow {
+
+namespace {
+void defaultTensorInitializer(Function *F, Node *node, unsigned inputIdx,
+                              Tensor *tensor) {
+  switch (node->getKind()) {
+  case Kinded::Kind::ConvolutionNodeKind: {
+    ConvolutionNode *CN = llvm::dyn_cast<ConvolutionNode>(node);
+    if (inputIdx == CN->FilterIdx) {
+      ShapeNHWC idim = ShapeNHWC(CN->getInput().dims());
+      ShapeHW kdim(CN->getKernels());
+      size_t fanIn = kdim.height * kdim.width * idim.c;
+      tensor->init(Tensor::InitKind::Xavier, fanIn, F->getPRNG());
+    } else if (inputIdx == CN->BiasIdx) {
+      tensor->init(Tensor::InitKind::Broadcast, 0.1, F->getPRNG());
+    }
+    break;
+  }
+  case Kinded::Kind::BatchNormalizationNodeKind: {
+    BatchNormalizationNode *BN = llvm::dyn_cast<BatchNormalizationNode>(node);
+    if (inputIdx == BN->ScaleIdx) {
+      tensor->init(glow::Tensor::InitKind::Zero, 0, F->getPRNG());
+    } else if (inputIdx == BN->BiasIdx) {
+      tensor->init(glow::Tensor::InitKind::Broadcast, 0.1, F->getPRNG());
+    } else if (inputIdx == BN->MeanIdx) {
+      tensor->zero();
+    } else if (inputIdx == BN->VarIdx) {
+      tensor->init(glow::Tensor::InitKind::Broadcast, 1.0, F->getPRNG());
+    }
+    break;
+  }
+  case Kinded::Kind::FullyConnectedNodeKind: {
+    FullyConnectedNode *FCN = llvm::dyn_cast<FullyConnectedNode>(node);
+    if (inputIdx == FCN->WeightsIdx) {
+      auto in = FCN->getInput();
+      tensor->init(glow::Tensor::InitKind::Zero, in.dims()[1], F->getPRNG());
+    } else if (inputIdx == FCN->BiasIdx) {
+      tensor->init(glow::Tensor::InitKind::Broadcast, 0.1, F->getPRNG());
+    }
+    break;
+  }
+  default:
+    break;
+  }
+}
+} // namespace
+
+TensorInitializer getDefaultTensorInitializer() {
+  return defaultTensorInitializer;
+}
+
+llvm::Error prepareFunctionForTraining(Function *F,
+                                       PlaceholderBindings &bindings,
+                                       TensorInitializer &&initializer) {
+
+  auto &nodes = F->getNodes();
+
+  Node *outputNode{nullptr};
```

Comment:
Can you explain why we need to special case output nodes? When we have the loop below that actually does the Constant -> PH conversion, if we encounter a SaveNode we could skip it just like we do Storage Nodes, no? Or if its input is a Constant I suppose we could even convert its Placeholder to a Constant, though it's a bit of a silly case?

Reply:
If output node (SoftMax) constant(s) got replaced with Placeholder then verification fails. Skipping SaveNode doesn't help, cause we need actually skip "SoftMax" of similar aggregation output node, which is in the chain: SaveNode -> Input (ReshapeNode) -> Input (SoftMax or similar).

Reply:
I'm still thinking this diff idea won't work in general case. Having one place where we load graph and making proper replacements (in a training mode) is a more reliable approach.

Reply:
I don't understand -- why is it a problem to convert the Softmax's `Selected` `Constant` input to a `Placeholder`? What verification fails?

Reply:
Here is the errors output:
Loading resnet50 model.
Creating input/output placeholders.
Preparing for training.
Differentiating graph.
Compiling backend.
Unsupported node found while compiling Function resnet50_grad for backend Interpreter: Add
name : gij_with_l2214
LHS : index64<1 x 1>
RHS : index64<1 x 1>
users : 1
Result : index64<1 x 1>
Unsupported node found while compiling Function resnet50_grad for backend Interpreter: Add
name : dx_with_momentum214
LHS : index64<1 x 1>
RHS : index64<1 x 1>
users : 2
Result : index64<1 x 1>
Unsupported node found while compiling Function resnet50_grad for backend Interpreter: Add
name : newW214
LHS : index64<1 x 1>
RHS : index64<1 x 1>
users : 1
Result : index64<1 x 1>
Unsupported node found while compiling Function resnet50_grad for backend Interpreter: Mul
name : L2Grad2141
LHS : index64<1 x 1>
RHS : index64<1 x 1>
users : 1
Result : index64<1 x 1>
Unsupported node found while compiling Function resnet50_grad for backend Interpreter: Mul
name : dx2141
LHS : index64<1 x 1>
RHS : index64<1 x 1>
users : 1
Result : index64<1 x 1>
Unsupported node found while compiling Function resnet50_grad for backend Interpreter: Mul
name : GsumMult2141
LHS : index64<1 x 1>
RHS : index64<1 x 1>
users : 1
Result : index64<1 x 1>
Encountered an error, exiting.
location: /Users/putivsky/local/pytorch/glow/lib/Optimizer/GraphOptimizer.cpp:2725 error code: COMPILE_UNSUPPORTED_NODE_AFTER_OPTIMIZE message: Unsupported node(s) found after optimizing Function resnet50_grad for backend Interpreter
p